nemo/collections/asr/parts/preprocessing/segment.py

```diff
@@ -186,19 +187,27 @@ def segment_from_file(cls, audio_file, target_sr=None, n_segments=0, trim=False,
         try:
             with sf.SoundFile(audio_file, 'r') as f:
                 sample_rate = f.samplerate
-                if 0 < n_segments < len(f):
-                    max_audio_start = len(f) - n_segments
+                if target_sr is not None:
+                    n_segments_at_original_sr = math.ceil(n_segments * sample_rate / target_sr)
```

Comment:
As mentioned in-person--please double-check the test case for both upsampling and downsampling

Reply:
Out of curiosity, why do we have to handle the case that `sample_rate` mismatches `target_sr` in the func `sefment_from_file`? In the return `cls()`, resampling would be applied as shown in https://github.com/NVIDIA/NeMo/blob/main/nemo/collections/asr/parts/preprocessing/segment.py#L98.

Reply:
IIRC Paarth was running into an issue where the mismatch caused an unexpected number of segments to be returned since `n_segments` calculated at the original sampling rate would not be the same as for the target sample rate. So this needed to be corrected for before passing in the segments for resampling.

Reply:
yes. I remembered that (see my above comment with detailed errors). But one thing that confused me: 
* why `n_segments (num of samples)` was independent to `sample_rate`, and why it was not automatically calculated based on `seconds * sample_rate`? 
* If you look at https://github.com/NVIDIA/NeMo/blob/main/examples/tts/conf/hifigan/hifigan_44100.yaml#L20, it applies for some fixed number as `train_n_segments: 16384`. Do we assume this number applies to any sample rate?

Reply:
Honestly, I'm not completely sure. But if I had to guess, maybe something to do with GPU memory during training? Similar to how we have a max audio length cutoff, perhaps...

Reply:
Yes, I think for HiFiGAN the train_n_segments should be 16384 for any sampling rate (Probably because it upsamples the spectrogram by some power of 2 at each layer). Also, I think the expected behavior of the function should be to get n_segments at the specified sampling rate.

Reply:
Thank you guys for the answers. It helps me understand well. Approved.